### PR TITLE
feat: Improve credential support for URL signing.

### LINKED ISSUE #8413: Improve how Cloud Storage presigned requests are generated
Currently, pre-signing requests for cloud storage is - for lack of better words - [a bit painful](https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Storage.V1/latest/index#signing-urls-without-a-service-account-credential-file).

I would love to see it as simple as:

```csharp
var credential = (await GoogleCredential.GetDefaultServiceAccount()) ?? throw new("No service account assigned.");
var urlSigner = await UrlSigner.FromServiceAccountCredential(credential);
var uri = await urlSigner.SignAsync(someBucket, key, duration, method);
```

The key focus here is that I think the APIs could afford to be a little more discoverable and self-evident in terms of how a user can obtain whatever credential type that satisfies their situation. The most crucial aspect being that right now, service accounts feel kind of second-class, which inhibits the convenience of being able to assign them to things like Google Cloud Run revisions . :slightly_smiling_face:

### PR DESCRIPTION
**Pull request by amanda-tarafa**
Closes #8413

### CODE REVIEW

**Comment by amanda-tarafa** [comment]
Draft for now, because snippet changes and possibly tests are missing.
Also the companion Google.Apis.Auth PR will be coming after, but that only has to make ComputeCredential an IBlobSigner.
This is mostly to agree on surface, etc.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs [hunk 1]
```diff
@@ -93,11 +161,8 @@ public static UrlSigner FromServiceAccountData(Stream credentialData)
         /// Creates a new <see cref="UrlSigner"/> instance for a service account.
         /// </summary>
         /// <param name="credential">The credential for the a service account. Must not be null.</param>
-        public static UrlSigner FromServiceAccountCredential(ServiceAccountCredential credential)
-        {
-            GaxPreconditions.CheckNotNull(credential, nameof(credential));
-            return new UrlSigner(new ServiceAccountCredentialBlobSigner(credential), SystemClock.Instance);
-        }
+        // TODO: Deprecate this method in favour of the new one.
```
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs:164]: Note that if you want to leave the TODO and deprecate it later, the comment will need to move.

**Review by jskeet — COMMENTED** [review]
Yup, this looks good.

**Comment by amanda-tarafa** [comment]
Requirement PR: https://github.com/googleapis/google-api-dotnet-client/pull/2308

**Comment by amanda-tarafa** [comment]
@jskeet this now includes all method deprecations and alternatives. PTAL just before I make any changes in tests due to deprecation, as those are a lot.
Also, some factory methods are now async given that the Compute service account ID fetch operation is async.

(Of course this PR won't build until we have merged and release https://github.com/googleapis/google-api-dotnet-client/pull/2308)

**Comment by amanda-tarafa** [comment]
And ignore the first commit marked `temp`.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs [hunk 2]
```diff
@@ -60,21 +62,124 @@ private UrlSigner(IBlobSigner blobSigner, IClock clock)
...
+            // Compute credentials are not specified through configuration files, and that's the only
+            // case in which this call may be async. It's safe to just return result here.
+            FromCredentialAsync(GoogleCredential.FromFile(credentialFilePath)).ResultWithUnwrappedExceptions();
+
+        /// <summary>
+        /// Creates a new <see cref="UrlSigner"/> instance from the JSON configuration file of a Google credential,
+        /// which is read asyncrhonously.
+        /// See <see cref="FromCredentialAsync(GoogleCredential, CancellationToken)"/> for more information about supported credential types.
+        /// </summary>
+        public static async Task<UrlSigner> FromCredentialFilePathAsync(string credentialFilePath, CancellationToken cancellationToken = default) =>
```
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs:150]: Possibly just FromCredentialFileAsync? That would be consistent with GoogleCredential.FromFileAsync.

I think it's fine for *this* method to be async, btw, given that the asynchronous part is loading the credential.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs [hunk 3]
```diff
@@ -60,21 +62,124 @@ private UrlSigner(IBlobSigner blobSigner, IClock clock)
...
+        /// Must wrap a credential currently supported for signing.
+        /// </param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <exception cref="InvalidOperationException">
+        /// The credential wrapped by the Google credential is not currently supported for signing.
+        /// </exception>
+        /// <remarks>
+        /// Please see the specific UrlSigner.FromCredential overloads for more information.
+        /// </remarks>
+        public static async Task<UrlSigner> FromCredentialAsync(GoogleCredential credential, CancellationToken cancellationToken = default) =>
```
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs:127]: Does this definitely need to be asynchronous? Can't we wait until the blob signing operation (which is already asynchronous) to get the target principal?

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs [hunk 4]
```diff
@@ -60,21 +62,124 @@ private UrlSigner(IBlobSigner blobSigner, IClock clock)
...
+        public static async Task<UrlSigner> FromCredentialFilePathAsync(string credentialFilePath, CancellationToken cancellationToken = default) =>
+            await FromCredentialAsync(
+                await GoogleCredential.FromFileAsync(credentialFilePath, cancellationToken).ConfigureAwait(false),
+                cancellationToken).ConfigureAwait(false);
+
+        /// <summary>
+        /// Creates a new <see cref="UrlSigner"/> instance from a Stream containing the JSON configuration file of a Google Credential.
+        /// See <see cref="FromCredentialAsync(GoogleCredential, CancellationToken)"/> for more information about supported credential types.
+        /// </summary>
+        public static UrlSigner FromCredentialData(Stream credentialData) =>
```
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs:159]: FromCredentialStream maybe? (Again, to be consistent with GoogleCredential.FromStream)

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.CredentialBlobSigner.cs [hunk 5]
```diff
@@ -12,33 +12,34 @@
...
         /// </summary>
-        private sealed class ServiceAccountCredentialBlobSigner : IBlobSigner
+        private sealed class CredentialBlobSigner : IBlobSigner
         {
-            private readonly ServiceAccountCredential _credential;
+            private readonly ISignerCredential _credential;
 
-            internal ServiceAccountCredentialBlobSigner(ServiceAccountCredential credential) =>
-                _credential = credential;
+            internal CredentialBlobSigner(ISignerCredential credential, string id, string algorithm) =>
```
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.CredentialBlobSigner.cs:31]: I think if this took a `Func<Task<string>>` as an asynchronous ID provider, we could avoid some factory methods being async.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs (hunk 4 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs]: Yep, done

**Review by amanda-tarafa — COMMENTED** [review]
Comments addressed and new approach implemented (save for tests).

@jskeet PTAL, second and third commits only.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs [hunk 6]
```diff
@@ -102,22 +102,28 @@ public static UrlSigner FromServiceAccountCredential(ServiceAccountCredential cr
...
+
+        /// <summary>
+        /// Creates a new <see cref="UrlSigner"/> instance for a custom blob signer obtained
+        /// from the given blob signer factory.
+        /// Note the factory will be only called once, and the returned blob signer will be reused
+        /// for subsequent signing operations with this URL signer.
+        /// </summary>
+        /// <param name="signerAsyncFactory">
+        /// The blob signer async factory to use. Must not be null.
+        /// Must not return a null blob signer</param>
```
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs:207]: For some reason I can't leave a suggestion, but: "When called, must not return a null reference or a task which completes with a null result."

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs [hunk 7]
```diff
@@ -102,22 +102,28 @@ public static UrlSigner FromServiceAccountCredential(ServiceAccountCredential cr
...
+        /// <returns>A new <see cref="UrlSigner"/> that will use a blob signer obtained from the specified factory function.</returns>
+        public static UrlSigner FromBlobSignerAsyncFactory(Func<Task<IBlobSigner>> signerAsyncFactory) =>
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs:209]: Yep, on both, changing the name to provider and making the method private.

**Review by amanda-tarafa — COMMENTED** [review]
All comments addressed, changes will be pushed shortly.

File: apis/Google.Cloud.Storage.V1/docs/index.md [hunk 8]
```diff
@@ -57,6 +57,26 @@ Or write-only access to put specific object content into a bucket:
 
 {{sample:UrlSigner.SignedURLPut}}
 
+### Supported credential types for URL signig
```
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/docs/index.md:60]: signig => signing

File: apis/Google.Cloud.Storage.V1/docs/index.md [hunk 9]
```diff
@@ -57,6 +57,26 @@ Or write-only access to put specific object content into a bucket:
...
+
+`Google.Apis.Auth.OAuth2.ServiceAccountCredential`, `Google.Apis.Auth.OAuth2.ComputeCredential`
+and `Google.Apis.Auth.OAuth2.ImpersonatedCredential` are all supported credentials from
+which you can build a `UrlSigner` by calling the appropiate `UrlSigner.FromCredential`
+method overload. `Google.Apis.Auth.OAuth2.GoogleCredential` is also supported as long as the
+underlying credentials is one of the supported specific types. The following example demonstrates
```
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/docs/index.md:66]: credentials => credential

File: apis/Google.Cloud.Storage.V1/docs/index.md [hunk 10]
```diff
@@ -57,6 +57,26 @@ Or write-only access to put specific object content into a bucket:
...
+how to explicitly use a Compute credential for URL signing.
+
+{{sample:UrlSigner.ComputeSignedURLGet}}
+
+When using a `UrlSigner` take care to choose the appropiate sync or async version of the signing
```
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/docs/index.md:71]: I think "take care" is probably a little too strong. How about "it's worth being aware of the underlying synchronous/asynchronous nature" or something similar. (And I'd expand "sync" to "synchronous" everywhere, although "async" is probably okay.)

**Review by jskeet — APPROVED** [review]
A few comments on the documentation, but that's all.

I'm slightly nervous about the use of Task.Run in the signer factory, but we can revisit that later if we want to (as everything is private).

**Comment by amanda-tarafa** [comment]
>I'm slightly nervous about the use of Task.Run in the signer factory, but we can revisit that later if we want to (as everything is private).

@jskeet see the last commit and let me know if that looks better. I haven't addressed the rest of your comments yet, I'll do that before merging if you are OK with this one.

**Review by jskeet — APPROVED** [review]
Happy to merge this (once doc niggles are done), but as discussed offline we're going to wait until next week to release, in case we think of better options.

**Pull request merged**